src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -14,131 +14,177 @@ namespace System.Buffers
-        internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, out ReadOnlyMemory<T> data, out SequencePosition next)
+        private bool TryGetBuffer(in SequencePosition position, out ReadOnlyMemory<T> memory, out SequencePosition next)
         {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
-            next = default;
-            object startObject = start.GetObject();
-            if (startObject != null)
-            {
-                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
-            }
-            else
-            {
-                type = SequenceType.Empty;
-            }
+            object positionObject = position.GetObject();
+            if (positionObject == null)
+                goto ReturnFalse;
 
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
-
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
-                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+            int positionIndex = GetIndex(position);
+            SequenceType type = GetSequenceType();
 
             if (type == SequenceType.MultiSegment)
             {
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-                ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
-
-                data = startSegment.Memory;
+                object endObject = _sequenceEnd.GetObject();
+                if (endObject == null) // Empty Segment
+                    goto EndPositionNotReached;
 
-                if (startSegment != endObject)
+                // End segment
+                if (positionObject == endObject)
                 {
-                    ReadOnlySequenceSegment<T> nextSegment = startSegment.Next;
+                    ReadOnlySequenceSegment<T> positionSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(positionObject);
+                    if (positionSegment == null)
+                        goto EndPositionNotReached;
+
+                    // Bounds check
+                    if (positionSegment == _sequenceStart.GetObject() && positionIndex < GetIndex(_sequenceStart))
+                        goto PositionOutOfRange;
 
-                    if (nextSegment == null && startSegment != endObject)
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                    int length = GetIndex(_sequenceEnd) - positionIndex;
+                    if (length < 0)
+                        goto PositionOutOfRange;
+
+                    memory = positionSegment.Memory.Slice(positionIndex, length);
+                    next = default;
+                    return true;
+                }
 
+                // Start or Middle segment
+                {
```

Comment:
Don't create blocks without flow control statements.

Reply:
Why? I want local variables.

Reply:
Using these blocks to create separate scopes makes the code harder to read and maintain (especially with local variable name re-use and gotos). This method is already tricky as is so I would try to avoid additional complexity. I would recommend simplifying it. Also, I wouldn't use the trusted flag either (those checks should always happen).

Reply:
Goto used to throw exception or exit. So it did not make method more complex to understand.